docs/proposals/diamond.md

```diff
@@ -0,0 +1,73 @@
+      ```/diamonds/{repo}/{id}/diamond.yaml```   
+1. Each split in a diamond generates a filelist that is stored in the versioned metadata bucket.
+   1. Each split is given a client assigned ID. Example split1,split2,split3.. or 1,2,3.. or a set of UUIDs.
+   2. On retries the split ID should be the same for a given split.
+      1. A node on failure can rerun the same command, as long as the split id is the same
+      the previous filelist contents will be superseded by the ones uploaded later.
+   3. command: ```datamon bundle diamon add --repo {repo} --id {ID} --split-id {client generated ID for each split in a diamond} --source {path}```
+   4. backend model: In the vmetadata bucket the following path will be created or rewritten.
+      ```/diamonds/{repo}/{id}/filelist-{split-id}-{ksuid}.yaml```. KSUID suffix generated at start of the 
+      add command allows for ordering of uploads from the same split.
+   5. This command can also be reused to have incremental uploads before a bundle is created.
+   6. The ordering of ```add``` from the same ```--split-id``` can be strongly ordered based on a design similar to WAL.
+      This is a nice to have addition and can be added later in the implementation.
+1. At the end the file lists are combined into a bundle with the commit message. 
+    1. The eventual bundle has a single namespace and if 2 nodes in the diamond have duplicate files, one will win. 
+    The other files including the winner will be stored in a special path in the bundle. (Proposal ./.{ID}/conflicts/{orig path}/file-suffix of client ID)
+    2. command: ```datamon bundle diamond commit --repo {repo} --id {id} --message {message} --label {label}```
+    3. Optionally a commit can be set to fail on conflict.
+        1. command: ```datamon bundle diamond commit --repo {repo} --id {id} --message {message} --label {label} --no-confict``` 
+    4. A file is written to mark the completion of a diamond and the corresponding bundle ID is written. On failure 2 bundles can be created in the repo.
+    for the same diamond but only one will win closing the bundle. ```/diamonds/repo/{id}/bundle.yaml```
+## Other commands
+ 
+```datamon bundle diamond list --repo {repo}```
+
+```datamon bundle diamond list-files --repo {repo} --id {id}```
+
+## Alternate proposal
+
+### Skip split-id
+The ```split-id``` that is client created can be skipped and replaced by a datamon internally generated KSUID. On retries
+a new id would be used. The final merge at the end will deal with ordering the KSUID and picking a winner if there are 
+conflicts. 
+
+Pros:
+
+1. Client does not need to create or manage ```split-id```
+2. Easy of use (similar to 1)
+
+Cons:
+1. Datamon will loose the mapping of client side logic for splits to actual additions to a diamond. ```split-id``` allows
+the grouping of which files originated from which split. The client can log the output of the command to capture the details.
```

Comment:
I believe skipping the client split-id generation is a good idea - this is a cleaner, more intuitive interface because the client will not need to be concerned with the inner workings of Datamon (`split-id`s to me seem like a tangential detail from the client's perspective - it is just concerned with adding (a series of) uploads and knowing that other nodes may be uploading to this bundle as well, without caring about what specific id it is within the group of nodes uploading to the bundle). This also makes sense because even in the original proposal, Datamon does not have a view of which uploads from a given node are retries and which are additions (since both cases generate a new KSUID for each upload), so Datamon would have to figure this out anyway by determining conflicts and resolving them for each node (correct me if I'm wrong). This alternate proposal just pushes that task fully out to Datamon so that it has to do this across all nodes. Now there may be a performance enhancement in Datamon having this mapping of which filelists came from which node, but that improvement would have to be quantified in order to determine that having this mapping really is advantageous.

Reply:
@kerneltime this indeed reflects well our conversation and the ideas I pushed forward.
I am also in favor of removing the need for client to manage the split-id at the (reasonable) cost of adding some merge logic to handle incomplete `split add` commands.

Reply:
Updated